kernel/src/cpu/irq_state.rs

```diff
@@ -0,0 +1,242 @@
+    !irqs_enabled()
+}
+
+/// This structure keeps track of PerCpu IRQ states. It tracks the original IRQ
+/// state and how deep IRQ-disable calls have been nested. The use of atomics
+/// is necessary for interior mutability and to make state modifications safe
+/// wrt. to IRQs.
+///
+/// The original state needs to be stored to not accidentially enable IRQs in
+/// contexts which have IRQs disabled by other means, e.g. in an exception or
+/// NMI/HV context.
+#[derive(Debug, Default)]
+pub struct IrqState {
+    /// IRQ state when count was `0`
+    state: AtomicBool,
+    /// Depth of IRQ-disabled nesting
+    count: AtomicIsize,
+}
+
+impl IrqState {
+    /// Create a new instance of `IrqState`
+    pub fn new() -> Self {
+        Self {
+            state: AtomicBool::new(false),
+            count: AtomicIsize::new(0),
+        }
+    }
+
+    /// Increase IRQ-disable nesting level by 1. The method will disable IRQs.
+    ///
+    /// # Safety
+    ///
+    /// The caller needs to make sure to match the number of `disable` calls
+    /// with the number of `enable` calls.
+    #[inline(always)]
+    pub unsafe fn disable(&self) {
+        let state = irqs_enabled();
+
+        raw_irqs_disable();
+        let val = self.count.fetch_add(1, Ordering::Relaxed);
+
+        assert!(val >= 0);
+
+        if val == 0 {
+            self.state.store(state, Ordering::Relaxed)
+        }
+    }
+
+    /// Decrease IRQ-disable nesting level by 1. The method will restore the
+    /// original IRQ state when the nesting level reaches 0.
+    ///
+    /// # Safety
+    ///
+    /// The caller needs to make sure to match the number of `disable` calls
+    /// with the number of `enable` calls.
+    #[inline(always)]
+    pub unsafe fn enable(&self) {
+        debug_assert!(irqs_disabled());
+
+        self.count.fetch_sub(1, Ordering::Relaxed);
+        let val = self.count.load(Ordering::Relaxed);
+
+        assert!(val >= 0);
+
+        if val == 0 {
+            let state = self.state.load(Ordering::Relaxed);
+            if state {
+                raw_irqs_enable();
+            }
+        }
+    }
+
+    /// Returns the current nesting count
+    ///
+    /// # Returns
+    ///
+    /// Levels of IRQ-disable nesting currently active
+    pub fn count(&self) -> isize {
+        self.count.load(Ordering::Relaxed)
+    }
+}
+
+impl Drop for IrqState {
+    /// This struct should never be dropped. Add a debug check in case it is
+    /// dropped anyway.
+    fn drop(&mut self) {
+        let count = self.count.load(Ordering::Relaxed);
+        assert!(count == 0);
+    }
+}
+
+/// And IRQ guard which saves the current IRQ state and disabled interrupts
+/// upon creation. When the guard goes out of scope the previous IRQ state is
+/// restored.
+///
+/// The struct implements the `Default` and `Drop` traits for easy use.
+#[derive(Debug)]
+#[must_use = "if unused previous IRQ state will be immediatly restored"]
+pub struct IrqGuard;
```

Comment:
```suggestion
pub struct IrqGuard(());
```

Adding a private unit field makes it impossible to create an instance without calling `IrqGuard::new()` (or `IrqGuard::default()`).

Reply:
Thanks, learned something new again :)